### PR TITLE
[csharp] Address MaxTimeout obsolete build warning

### LINKED ISSUE #19525: [BUG] [csharp] Obsolete build warnings after PR to update to RestSharp 112.0.0
#### Bug Report Checklist

- [ ] Have you provided a full/minimal spec to reproduce the issue?
- [ ] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [ ] Have you [tested with the latest master](https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-to-test-with-the-latest-master-of-openapi-generator) to confirm the issue still exists?
- [x] Have you searched for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Sponsorship to speed up the bug fix or feature request ([example](https://github.com/OpenAPITools/openapi-generator/issues/6178))

##### Description

Due to vulnerabilities in the current generated RestSharp version

- #19387

I tried to update 7.8.0 generated C# client to RestSharp 112.0.0 as shown in

- #19507

But this produced two build warnings. For projects that treat warnings as errors in CI, this will fail the build.

```
ApiClient.cs(400,17): warning CS0618: 'RestClientOptions.MaxTimeout' is obsolete: 'Use Timeout instead.'
ApiClient.cs(472,24): warning CS0618: 'RestClientExtensions.Deserialize<T>(IRestClient, RestResponse)' is obsolete: 'Please use the async overload with a cancellation token'
```

The related upstream changes in RestSharp:

- https://github.com/restsharp/RestSharp/pull/2228
- https://github.com/restsharp/RestSharp/pull/2078
- https://github.com/restsharp/RestSharp/pull/2141

Here are where these obsolete methods are still used within `openapi-generator` `main` branch:

https://github.com/OpenAPITools/openapi-generator/blob/e914c4098b44ac734d3dd21300d433f6cb2974a5/modules/openapi-generator/src/main/resources/csharp/ApiClient.mustache#L458

https://github.com/OpenAPITools/openapi-generator/blob/e914c4098b44ac734d3dd21300d433f6cb2974a5/modules/openapi-generator/src/main/resources/csharp/ApiClient.mustache#L557

##### openapi-generator version

7.8.0 (with #19507 backport manually applied)

### PR DESCRIPTION
**Pull request by alec-petersen**
@wing328 , this adds that option to enable use of MaxTimeout over Timeout, but it defaults to the non-obsolete option by default. Fix #19525.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming `7.x.0` minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [ ] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/csharp/ApiClient.mustache [hunk 1]
```diff
@@ -455,7 +455,12 @@ namespace {{packageName}}.Client
             var clientOptions = new RestClientOptions(baseUrl)
             {
                 ClientCertificates = configuration.ClientCertificates,
+                {{#useMaxTimeout}}
                 MaxTimeout = configuration.Timeout,
+                {{/useMaxTimeout}}
+                {{^useMaxTimeout}}
+                Timeout = TimeSpan.FromMilliseconds(configuration.Timeout),
```
Comment (alec-petersen) [inline modules/openapi-generator/src/main/resources/csharp/ApiClient.mustache:462]: Open to making the configuration property in IReadableConfiguration a TimeSpan

Comment (wing328) [reply modules/openapi-generator/src/main/resources/csharp/ApiClient.mustache:462]: to me, it make senses and it's another breaking change with fallback so we should include it in this PR

**Comment by wing328** [comment]
tests passed via https://ci.appveyor.com/project/WilliamCheng/openapi-generator/builds/50804786

File: samples/client/petstore/csharp/restsharp/net4.7/MultipleFrameworks/src/Org.OpenAPITools/Client/Configuration.cs [hunk 2]
```diff
@@ -216,9 +216,9 @@ public virtual IDictionary<string, string> DefaultHeader
         public virtual IDictionary<string, string> DefaultHeaders { get; set; }
 
         /// <summary>
-        /// Gets or sets the HTTP timeout (milliseconds) of ApiClient. Default to 100000 milliseconds.
+        /// Gets or sets the HTTP timeout of ApiClient. Defaults to 100 seconds.
         /// </summary>
-        public virtual int Timeout { get; set; }
+        public virtual TimeSpan Timeout { get; set; }
```
Comment (wing328) [inline samples/client/petstore/csharp/restsharp/net4.7/MultipleFrameworks/src/Org.OpenAPITools/Client/Configuration.cs:221]: if users call the getter of Timeout in their code, they will need to update their code accordingly updating the C# client to a newer version including this change, right?

Comment (alec-petersen) [reply samples/client/petstore/csharp/restsharp/net4.7/MultipleFrameworks/src/Org.OpenAPITools/Client/Configuration.cs:221]: Correct, if they use the 7.10.0 version of the generator to generate their client, they will need to update usages of Timeout.

**Comment by wing328** [comment]
Added an option `useIntForTimeout` for fallback via https://github.com/OpenAPITools/openapi-generator/pull/20069, e.g. `--addititional-properties useIntForTimeout=true` in CLI

**Pull request merged**